Support binary_sensor and device_tracker in HomeKit.

Yonsm opened the pull request:
## Description:

Map binary_sensor and device_tracker to HomeKit Binary Sensor, based on 'device_class' attributes:

	'co2': 'CarbonDioxideSensor'
	'gas': 'CarbonMonoxideSensor'
	'moisture': 'LeakSensor'
	'motion': 'MotionSensor'
	'occupancy': 'OccupancySensor'
	'opening': 'ContactSensor'
	'smoke': 'SmokeSensor'

If there is no any matched device_class, e.g. device_tracker, then default sensor type is OccupancySensor.


## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`.

---

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5143

Commit:
Support binary_sensor and device_tracker for HomeKit

Commit:
Add test for get_accessory and binary sensor

Passed:
 tox -e py36 -- tests/components/homekit/test_type_sensors.py -x
tox -e py36 -- tests/components/homekit/test_get_accessories.py -x

Commit:
Wrap line to 79 chars

cdce8p commented on tests/components/homekit/test_type_sensors.py:
```diff
@@ -68,3 +69,39 @@ def test_humidity(self):
+
+        self.hass.states.set(entity_id, STATE_UNKNOWN,
+                             {ATTR_DEVICE_CLASS: "opening"})
+        self.hass.block_till_done()  # Ensure state.attributes
```
You don't need that comment.

Yonsm replied:
done

cdce8p reviewed (changes requested):
It's looking good so far. Unfortunately I can't test it until tomorrow evening (CET).
Since you dynamically assign the service and char, can you add tests to verify that the right ones where selected. You can get the service / char name with `service.display_name` or `char.display_name`. Those must be equal to the HomeKit char / service constants.

Commit:
Test service.display_name and char_detected.display_name

1. Passed: tox -e py36 -- tests/components/homekit/test_type_sensors.py
-x
2. And remove ‘# Ensure state.attributes’ comment

Commit:
Fix continuation line under-indented for visual indent

cdce8p commented on tests/components/homekit/test_type_sensors.py:
```diff
@@ -68,3 +70,44 @@ def test_humidity(self):
+        self.hass.states.set(entity_id, STATE_NOT_HOME,
+                             {ATTR_DEVICE_CLASS: "opening"})
+        self.hass.block_till_done()
+        self.assertEqual(acc.char_detected.value, 0)
```
Please add this to the end. That way you improve the coverage.
```py
        self.hass.states.remove(entity_id)
        self.hass.block_till_done()
```

cdce8p commented on tests/components/homekit/test_type_sensors.py:
```diff
@@ -68,3 +70,44 @@ def test_humidity(self):
+        self.assertEqual(acc.get_service(SERV_CONTACT_SENSOR).display_name,
+                         SERV_CONTACT_SENSOR)
+        self.assertEqual(acc.char_detected.display_name,
+                         CHAR_CONTACT_SENSOR_STATE)
```
Instead of L85 - L88, why not do something like that:
```py
from homeassistant.components.homekit.type_sensors import BINARY_SENSOR_SERVICE_MAP
from homeassistant.components.homekit.const import (
    DEVICE_CLASS_CO2, SERV_CARBON_DIOXIDE_SENSOR, CHAR_CARBON_DIOXIDE_DETECTED,
    DEVICE_CLASS_GAS, SERV_CARBON_MONOXIDE_SENSOR,
    CHAR_CARBON_MONOXIDE_DETECTED,
    DEVICE_CLASS_MOISTURE, SERV_LEAK_SENSOR, CHAR_LEAK_DETECTED,
    DEVICE_CLASS_MOTION, SERV_MOTION_SENSOR, CHAR_MOTION_DETECTED,
    DEVICE_CLASS_OCCUPANCY, SERV_OCCUPANCY_SENSOR, CHAR_OCCUPANCY_DETECTED,
    DEVICE_CLASS_OPENING, SERV_CONTACT_SENSOR, CHAR_CONTACT_SENSOR_STATE,
    DEVICE_CLASS_SMOKE, SERV_SMOKE_SENSOR, CHAR_SMOKE_DETECTED)

class TestHomekitSensors(unittest.TestCase):

#####
#####

    def test_binary(self):

#####
#####

    def test_binary_device_classes(self):

        entity_id = 'binary_sensor.demo'

        for device_class, (service, char) in BINARY_SENSOR_SERVICE_MAP.items():
            print(device_class)
            self.hass.states.set(entity_id, STATE_OFF,
                                 {ATTR_DEVICE_CLASS: device_class})
            self.hass.block_till_done()

            acc = BinarySensor(self.hass, entity_id, 'Binary Sensor', aid=2)
            self.assertEqual(acc.get_service(service).display_name, service)
            self.assertEqual(acc.char_detected.display_name, char)

```
That way you check if every service and char can be added successfully.

Yonsm replied:
done

Commit:
Refine testcase for type_sensors.py with more coverage

PASSED: tox -e py36 -- tests/components/homekit/test_type_sensors.py
PASSED: script/lint (in my submit)

cdce8p committed:
Split test to improve speed

cdce8p commented:
Thanks :+1:
Looking forward to more PR's

Yonsm commented:
Thank you for your effective advice and guidance.

shuaiger commented:
cong~ merged!

nicx commented:
@Yonsm great, thanks a lot! @cdce8p maybe its possible to add a default mapping to binary_sensors with device_class "opening" for all sensors which returns states like "open/tilted/closed"? something similar to the existing temperature sensors. with that change I don't need to create extra template binary_sensors anymore :)

nicx commented:
@cdce8p great to hear, I can wait for this 👍 keep up this great work! 🥇

spacesuitdiver commented:
Or maybe it's best to default to `ContactSensor` since this allows the most flexibility in Homekit and is probably the likely use of a binary_sensor?

MartinHjelmare commented:
Please don't discuss feature requests in closed PRs.

spacesuitdiver commented:
@MartinHjelmare seemed like an omission, what SHOULD I do instead?

MartinHjelmare commented:
If you want to suggest an enhancement please open a feature request in [the Feature Requests section](https://community.home-assistant.io/c/feature-requests) of our community forum.

Merged PRs should not be used for enhancement discussion or bug reports. If you've found a bug it's ok to make a review with inline comments and link to an issue that reports the bug.

Thanks!
